core/course_controller/CourseController.py

```diff
@@ -1,48 +1,226 @@
+        lab_sections = []
+
+        # Retrieve course sections
+        for section in CourseSection.objects.filter(course=course):
+            section_instructor = UserRef(
+                name=f"{section.instructor.first_name} {section.instructor.last_name}",
+                username=section.instructor.username,
+            )
+            course_sections.append(CourseSectionRef(
+                section_number=str(section.course_section_number),
+                instructor=section_instructor,
+            ))
+
+        # Retrieve lab sections
+        for lab in LabSection.objects.filter(course=course):
+            ta = lab.get_ta()
+            ta_ref = UserRef(
+                name=f"{ta.first_name} {ta.last_name}",
+                username=ta.username,
+            ) if ta else None
+            lab_sections.append(LabSectionRef(
+                section_number=str(lab.lab_section_number),
+                instructor=ta_ref,
+            ))
+
+        return CourseOverview(
+            code=course.course_code,
+            name=course.course_name,
+            instructor=instructor,
+            course_sections=course_sections,
+            lab_sections=lab_sections,
+        )
 
     @staticmethod
-    def search_courses(course_search: str, semester_id: int | None = None) -> List[CourseRef]:
+    def search_courses(course_search: str, semester_name: str | None = None) -> List[CourseRef]:
         """
         Pre-conditions: Semester is a valid value if given
         Post-conditions: Returns a list of courses whose title or code
             matches the course_search_str.
         Side-effects: N/A
         """
-        pass
+        results = Course.objects.all()
+
+        if semester_name:
+            results = results.filter(semester__semester_name=semester_name)
+
+        if course_search:
+            results = results.filter(
+                models.Q(course_name__icontains=course_search) |
+                models.Q(course_code__icontains=course_search)
+            )
+
+        return [
+            CourseRef(course_code=course.course_code, course_name=course.course_name)
+            for course in results
+        ]
     
     @staticmethod
-    def delete_course(course_id: int) -> None:
+    def delete_course(course_code: str) -> None:
         """
         Pre-conditions: Course id is a valid value matching a record in courses.
         Post-conditions: Removes a record from Course with matching course_id.
-        Side-effects: Adds new record to Course table.
+        Side-effects: removes matching record from Course table and any objects with foreign key references to it.
+        """
+        try:
+            course = Course.objects.get(course_code=course_code)
+            course.delete()
+        except Course.DoesNotExist:
+            raise ValueError("Course with the given code does not exist.")
+
+    @staticmethod
+    def save_lab_sections(course_data: CourseFormData) -> str:
```

Comment:
I was curious about private methods in python, and you can kind of do it by prefixing the method name with "__"

https://www.geeksforgeeks.org/private-methods-in-python/#

same goes for other helper method